editor/components/rich-text/index.js

```diff
@@ -789,12 +789,33 @@ export class RichText extends Component {
 						return;
 					}
 
-					const anchor = this.editor.dom.getParent( this.editor.selection.getNode(), 'a' );
-					if ( ! anchor ) {
-						this.removeFormat( 'link' );
-					}
-					const { value: href, ...params } = formatValue;
-					this.applyFormat( 'link', { href, ...params }, anchor );
+					let anchor;
+
+					// Create only one undo level when inserting and modifying the link
+					this.editor.undoManager.transact( () => {
+						// Use built-in TinyMCE command to insert the link. This takes care of
+						// deleting any existing links within the selection
+						const { value: href, ...params } = formatValue;
+						this.editor.execCommand( 'mceInsertLink', false, {
+							href,
+							...params,
+							'data-wp-temp-link': 1,
+						} );
+
+						// mceInsertLink doesn't return the anchor it creates, so query for it
+						// using a temporary attribute
+						anchor = this.editor.getBody().querySelector( '[data-wp-temp-link]' );
+						anchor.removeAttribute( 'data-wp-temp-link' );
```

Comment:
I'm not quite following why it is necessary to use the temporary attribute. Can't we insert the text and element at once? So I'm guessing when the selection is collapsed, we could use `editor.insertContent( '<a href="link">link</a>' )` or something like that?

Reply:
I wasn't having any luck with `isCollapsed()` yesterday but now when I try it it's fine—amazing what some sleep can do! I've reworked this in 4a02305ff1472d58a32e14fca6abbb01b0fea827 🙂